src/spinnaker.cpp

```diff
@@ -0,0 +1,1060 @@
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+    return Device_Err;
+}
+
+enum DeviceStatusCode
+spinnakercam_close(struct Driver* self_, struct Device* in)
+{
+    try {
+        CHECK(self_);
+        ((struct SpinnakerDriver*)self_)->close(in);
+        return Device_Ok;
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+    return Device_Err;
+}
+
+enum DeviceStatusCode
+spinnakercam_shutdown(struct Driver* self_)
+{
+    try {
+        CHECK(self_);
+        delete (struct SpinnakerDriver*)self_;
+        return Device_Ok;
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+    return Device_Err;
+}
+
+//
+// Driver implementation
+//
+
+SpinnakerDriver::SpinnakerDriver()
+  : Driver{
+      .device_count = ::spinnakercam_device_count,
+      .describe = ::spinnakercam_describe,
+      .open = ::spinnakercam_open,
+      .close = ::spinnakercam_close,
+      .shutdown = ::spinnakercam_shutdown,
+  },
+  system_(Spinnaker::System::GetInstance())
+{
+}
+
+SpinnakerDriver::~SpinnakerDriver()
+{
+    try {
+        shutdown();
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+}
+
+void
+SpinnakerDriver::describe(DeviceIdentifier* identifier, uint64_t i)
+{
+    check_spinnaker_camera_id(i);
+
+    Spinnaker::CameraList camera_list = system_->GetCameras();
+    Spinnaker::CameraPtr camera = camera_list.GetByIndex((unsigned int)i);
+    Spinnaker::GenApi::INodeMap& nodeMap = camera->GetTLDeviceNodeMap();
+
+    const Spinnaker::GenApi::CStringPtr vendor_name =
+      nodeMap.GetNode("DeviceVendorName");
+    const Spinnaker::GenApi::CStringPtr device_name =
+      nodeMap.GetNode("DeviceModelName");
+    const Spinnaker::GenApi::CStringPtr device_sn =
+      nodeMap.GetNode("DeviceSerialNumber");
+
+    *identifier = DeviceIdentifier{
+        .device_id = (uint8_t)i,
+        .kind = DeviceKind_Camera,
+    };
+
+    snprintf(identifier->name,
+             sizeof(identifier->name),
+             "%s %s %s",
+             vendor_name->GetValue().c_str(),
+             device_name->GetValue().c_str(),
+             device_sn->GetValue().c_str());
+}
+
+uint32_t
+SpinnakerDriver::device_count()
+{
+    return (uint32_t)system_->GetCameras().GetSize();
+}
+
+void
+SpinnakerDriver::open(uint64_t device_id, struct Device** out)
+{
+    CHECK(out);
+    check_spinnaker_camera_id(device_id);
+    Spinnaker::CameraList camera_list = system_->GetCameras();
+    Spinnaker::CameraPtr camera =
+      camera_list.GetByIndex((unsigned int)device_id);
+    // TODO: SpinnakerCamera may partially initialize the camera
+    // and throw, so we may want a try/catch block here that deinits
+    // the camera if it was initialized.
```

Comment:
When do you think you'll revisit this?

Reply:
I'm not sure, partially because I'm not sure where this logic should live - here, in SpinnakerCamera's constructor, or in the acquire video runtime.

In general, the remaining TODOs are ones that I could not resolve on my own, so any input would be much appreciated.

Would be great to have no remaining TODOs before merging. Or if there are some, they should be tracked by GitHub issues.

Reply:
This is roughly analogous to the camera restart logic we (mostly Nathan) wrote into `dcam_start`. I think it makes sense to live here, but unless you're actually seeing the problem (and / or can reproduce it in a test) I would say it's fine to leave it for now as long as you make an issue.

Reply:
Moved it from here to the constructor. That didn't seem to cause any new problems, but I'll verify tomorrow.